lib/network/kmsAWS/Client.ts

```diff
@@ -0,0 +1,256 @@
+        const params = {};
+        this.client.createKey(params, (err: AWSError, data) => {
+            if (err) {
+                const error = _arsenalError(err);
+                logger.error("AWS KMS: failed to create master encryption key", { err });
+                cb(error);
+                return;
+            }
+            logger.debug("AWS KMS: master encryption key created", { KeyMetadata: data?.KeyMetadata });
+            cb(null, data?.KeyMetadata?.KeyId);
+        });
+    }
+
+    destroyBucketKey(bucketKeyId: string, logger: werelogs.Logger, cb: (err: Error | null) => void): void {
+        logger.debug("AWS KMS: deleting encryption key managed at the bucket level", { bucketKeyId });
+        this.deleteMasterKey(bucketKeyId, logger, cb);
+    }
+
+    deleteMasterKey(masterKeyId: string, logger: werelogs.Logger, cb: (err: Error | null) => void): void {
+        logger.debug("AWS KMS: deleting master encryption key", { masterKeyId });
+        const params = {
+            KeyId: masterKeyId,
+            PendingWindowInDays: 7,
+        };
+        this.client.scheduleKeyDeletion(params, (err: AWSError, data) => {
+            if (err) {
+                const error = _arsenalError(err);
+                logger.error("AWS KMS: failed to delete master encryption key", { err });
+                cb(error);
+                return;
+            }
+
+            if (data?.KeyState && data.KeyState !== "PendingDeletion") {
+                const error = _arsenalError("key is not in PendingDeletion state");
+                logger.error("AWS KMS: failed to delete master encryption key", { err, data });
+                cb(error);
+                return;
+            }
+
+            cb(null);
+        });
+    }
+
+    generateDataKey(
+        cryptoScheme: number,
+        masterKeyId: string,
+        logger: werelogs.Logger,
+        cb: (err: Error | null, plainTextDataKey?: Buffer, cipheredDataKey?: Buffer) => void
+    ): void {
+        logger.debug("AWS KMS: generating data key", { cryptoScheme, masterKeyId });
+        assert.strictEqual(cryptoScheme, 1);
+
+        const params = {
+            KeyId: masterKeyId,
+            KeySpec: 'AES_256',
+        };
+
+        this.client.generateDataKey(params, (err: AWSError, data) => {
+            if (err) {
+                const error = _arsenalError(err);
+                logger.error("AWS KMS: failed to generate data key", { err });
+                cb(error);
+                return;
+            }
+
+            if (!data) {
+                const error = _arsenalError("failed to generate data key: empty response");
+                logger.error("AWS KMS: failed to generate data key: empty response");
+                cb(error);
+                return;
+            }
+
+            logger.debug("AWS KMS: data key generated");
+            cb(null, Buffer.from(data.Plaintext as Uint8Array), Buffer.from(data.CiphertextBlob as Uint8Array));
```

Comment:
For `generateDataKey` and `decrypt`, the documentation says:

> Then erase the plaintext data key from memory.

https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateDataKey.html

We should probably clear the sensitive key from the received buffe like it's done herer: https://github.com/aws/aws-encryption-sdk-javascript/blob/e248a96ef02e51baa00e1b0d71952406bd4cf78c/modules/kms-keyring/src/helpers.ts#L159-L173

Reply:
We already zero out the plaintext key after [generateDataKey](https://github.com/scality/cloudserver/blob/development/7.70/lib/kms/wrapper.js#L204) and [decrypt](https://github.com/scality/cloudserver/blob/development/7.70/lib/kms/wrapper.js#L279). However, this approach is not sufficient because if the buffer is accessed or copied before fill(0) is called the plaintext key could be temporarily exposed in memory. For that reason, we introduced the `safePlaintext` method.

Reply:
Note that `Buffer.from` uses a nodejs shared buffer pool for perf reason for small (<4Kb) Buffer.

Any other places in the process can access memory allocated this way by doing `Buffer.from('1').buffer # Access to the internal 8Kb ArrayBuffer pool`.

If you want to not use that shared pool you can use `Buffer.alloc` or `Buffer.allocUnsafeSlow` and copy the buffer.

Reply:
updated there: https://github.com/scality/Arsenal/compare/5ec3eeec9b0fb0da4f6fc46241507fcdc27a1745..5d412c20a3e93bc2dfc69ee95a169e03386b2035

Reply:
`Buffer.alloc` is slower than `Buffer.allocUnsafeSlow` because it fills the buffer with `0`.

The naming is weird but:
- `Buffer.alloc`: alloc outside the shared pool, initialize with `0`.
- `Buffer.allocUnsafe`: alloc from the shared pool, not initialized.
- `Buffer.allocUnsafeSlow`: alloc outside the shared pool, not initialized.

You can pass the data to `Buffer.alloc` second param to fill it directly with data without filling with `0` before. (It's like `Buffer.allocUnsafeSlow`)

Reply:
Good catch! Sorry for the back and forth.

Reply:
Latest update: https://github.com/scality/Arsenal/compare/6a74794998e635b52c2b48abca9735d66a336404..82a3420c5c59b68924c3968970cc73e755ccbb59